components/script/dom/webglrenderingcontext.rs

```diff
@@ -3805,48 +3805,50 @@ impl UniformSetterType {
     }
 }
 
-#[derive(Default, JSTraceable, MallocSizeOf)]
+#[derive(JSTraceable, MallocSizeOf)]
 #[must_root]
-pub struct BoundAttribBuffers {
-    elements: DomRefCell<FnvHashMap<u32, (bool, Option<Dom<WebGLBuffer>>)>>,
+pub struct VertexAttribs {
+    attribs: DomRefCell<Box<[(bool, Option<Dom<WebGLBuffer>>)]>>,
 }
 
-impl BoundAttribBuffers {
+impl VertexAttribs {
+    pub fn new(max: u32) -> Self {
+        Self { attribs: DomRefCell::new(vec![Default::default(); max as usize].into()) }
+    }
+
     pub fn clear(&self) {
-        self.elements.borrow_mut().clear()
+        for attrib in &mut **self.attribs.borrow_mut() {
+            *attrib = Default::default();
+        }
     }
 
-    pub fn set_from(&self, other: &BoundAttribBuffers) {
-        *self.elements.borrow_mut() = other.elements.borrow().clone();
+    pub fn set_from(&self, other: &Self) {
+        self.attribs.borrow_mut().clone_from_slice(&other.attribs.borrow());
     }
 
-    pub fn borrow(&self) -> Ref<FnvHashMap<u32, (bool, Option<Dom<WebGLBuffer>>)>> {
-        self.elements.borrow()
+    pub fn borrow(&self) -> Ref<[(bool, Option<Dom<WebGLBuffer>>)]> {
+        Ref::map(self.attribs.borrow(), |attribs| &**attribs)
     }
 
-    fn remove_buffer(&self, buffer: &WebGLBuffer) {
-        self.elements.borrow_mut().retain(|_, v| {
-            v.1.as_ref().map_or(true, |b| b.id() != buffer.id())
-        })
+    fn delete_buffer(&self, buffer: &WebGLBuffer) {
+        for attrib in &mut **self.attribs.borrow_mut() {
+            if attrib.1.as_ref().map_or(false, |b| b.id() == buffer.id()) {
+                attrib.1 = None;
+            }
+        }
     }
 
     fn get(&self, index: u32) -> Option<Ref<WebGLBuffer>> {
```

Comment:
A bunch of these should probably `debug_assert!(index as usize < self.elements.borrow().len(), "Someone forgot to validate properly")` or something of the sort, wdyt?